include/deal.II/base/mpi_compute_index_owner_internal.h

```diff
@@ -0,0 +1,799 @@
+                Assert(it != actually_owning_rank_list.end(),
+                       ExcInternalError());
+                Assert(*it == rank_in_owned_indices, ExcInternalError());
+                return it - actually_owning_rank_list.begin();
+              }
+          }
+
+        private:
+          /**
+           * Compute the partition from the global size of the index space and
+           * the number of ranks.
+           */
+          void
+          partition(const IndexSet &owned_indices, const MPI_Comm &comm)
+          {
+#ifdef DEAL_II_WITH_MPI
+            const unsigned int n_procs = n_mpi_processes(comm);
+            const unsigned int my_rank = this_mpi_process(comm);
+
+            size              = owned_indices.size();
+            dofs_per_process  = (size + n_procs - 1) / n_procs;
+            local_range.first = std::min(dofs_per_process * my_rank, size);
+            local_range.second =
+              std::min(dofs_per_process * (my_rank + 1), size);
+            local_size = local_range.second - local_range.first;
+
+            actually_owning_ranks.resize(local_size);
+#else
+            (void)owned_indices;
+            (void)comm;
+#endif
+          }
+        };
+
+
+
+        /**
+         * Specialization of ConsensusAlgorithmProcess for the context of
+         * Utilities::MPI::compute_index_owner() and
+         * Utilities::MPI::Partitioner::set_ghost_indices().
+         */
+        class ConsensusAlgorithmProcess
+          : public dealii::Utilities::MPI::
+              ConsensusAlgorithmProcess<types::global_dof_index, unsigned int>
+        {
+        public:
+          /**
+           * Constructor.
+           */
+          ConsensusAlgorithmProcess(const IndexSet &owned_indices,
+                                    const IndexSet &indices_to_look_up,
+                                    const MPI_Comm &comm,
+                                    std::vector<unsigned int> &owning_ranks,
+                                    const bool track_index_requests = false)
+            : owned_indices(owned_indices)
+            , indices_to_look_up(indices_to_look_up)
+            , comm(comm)
+            , my_rank(this_mpi_process(comm))
+            , n_procs(n_mpi_processes(comm))
+            , track_index_requests(track_index_requests)
+            , owning_ranks(owning_ranks)
+          {
+            dict.reinit(owned_indices, comm);
+            requesters.resize(dict.actually_owning_rank_list.size());
+          }
+
+          /**
+           * The index space which describes the locally owned space.
+           */
+          const IndexSet &owned_indices;
+
+          /**
+           * The indices which are "ghosts" on a given rank and should be
+           * looked up in terms of their owner rank from owned_indices.
+           */
+          const IndexSet &indices_to_look_up;
+
+          /**
+           * The underlying MPI communicator.
+           */
+          const MPI_Comm &comm;
```

Comment:
this is an int internally, right? I think we normally copy it.

Reply:
It's not specified. It could be implemented as a struct, or a pointer to something. It's better to copy it.